### ISSUE TITLE
Security related telemetry support in gRPC

### ISSUE THREAD

**Issue by ZhenLian** [grpc/grpc#25626]
It would be great if we can have a standard way in gRPC to gather telemetry metrics based on the authentication/authorization results. For example, it would be great if we can have:
1. counters for authentication success/failure
2. counters for authorization success/failure
3. other auditing information for authorization, such as peer identity name, etc

**Comment by sanjaypujare** [comment]
What are the use-cases for this? Also does it pose a security risk to expose security related metrics (e.g. authz failures) without additional guards?

**Comment by danielywong** [comment]
+1.  I also believe it is a useful feature for gRPC to (1) standardize metrics collection internally and (2) provide basic counters especially those applications have no access to.

For security, metrics such as authentication failures and unauthorized access attempts are currently not visible to the applications.  For example, these metrics can provide very useful security signals for mission critical systems with limited compute power.

sanjaypujare@: Do you have any specific security risk you have in mind?  Essentially we are just giving access to the local applications about external accesses to their services, right?

**Comment by sanjaypujare** [comment]
> sanjaypujare@: Do you have any specific security risk you have in mind? Essentially we are just giving access to the local applications about external accesses to their services, right?

No specific risk in mind. Just wondering about an attacker getting a hold of the metrics.

**Comment by markdroth** [comment]
This problem is actually not specific to security.  gRPC does not have any general-purpose mechanism for exporting *any* metrics in OSS.  If we had such a mechanism, we could use it for security and anything else, but it's not clear what such a mechanism would look like in the OSS world.

**Comment by danielywong** [comment]
Thanks, Mark and Sanjay.  "general-purpose mechanism for any metrics" definitely sounds like the way to go for long term. 
 Meanwhile, do you have any suggest for gRPC to provide applications visibility on authentication and authorization failures specifically?  Callback function?  Syslog option?  (yeah, I can see callback causing threading/blocking issue and syslog has privacy concerns)

**Comment by markdroth** [comment]
The only way we have to export this information today is in the RPC status error messages.  We could do a better job at that, as per #22883.  But even once that's fixed, it won't really help if what you're after is programmatic access to detailed metrics.

I'm sorry we don't have a solution to offer you here.

**Comment by danielywong** [comment]
Thanks again, Mark.  I totally understand this is a priority call.

Just to make sure I understand your reply ... even when #22883 got fixed, only client applications will get the improved status error while the server applications still have no knowledge of the rejected accesses, right?

**Comment by jiangtaoli2016** [comment]
OSS gRPC is a library. It is hard to provide generic logging. At best, it can provide some audit pluggin.
Take Istio example, there is [stackdriver plugin](https://istio.io/latest/docs/reference/config/proxy_extensions/stackdriver/) for auditing log. It is optional and only works for GCP. 

@markdroth We may consider add such optional audit plugin in the future similar to Envoy/Istio.

**Comment by markdroth** [comment]
@danielywong Correct, the RPC statuses are only for the client, not for the server.

@srini100 We should consider this use-case as we invest more in observability architecture.

**Comment by dfawley** [comment]
> gRPC does not have any general-purpose mechanism for exporting any metrics in OSS

Channelz?  This seems like something that could fit in there.

**Comment by markdroth** [comment]
It is possible that we could expose some of the channel-level metrics on channelz.  But I don't think we'd want to do that for call-level data, so this would cover (e.g.) mTLS handshake results but not call-level authz policy results.

Also, depending on the use-case, channelz may not be as easy to get the data out to other systems as something like OpenTelemetry.  I think we'll have to see how the ecosystem in this space develops.

**Comment by dfawley** [comment]
Agreed.  The three things listed in the initial comment are all pretty different things.  The last one isn't even a metric, per se.

I also wanted to mention: in Go and Java, some of these things may be feasible for the user to extract themselves right now by using a custom transport/channel creds plugin.

cc @ejona86

**Comment by ejona86** [comment]
1) In Java and Go, authz failures could be counted by the application via an interceptor since the authz code is an interceptor itself.
2) In Java, authn failures could be counted by the application by implementing a TrustManager that delegates to their real TrustManager, maybe using the Advanced TLS APIs. In Go, authn failures may be harder; it looks like you may need to use InsecureSkipVerify along with tls.Config.VerifyPeerCertificate. But that's not gRPC code anyway so it is a problem no matter what approach you want to take.
3) In Java, auditing information could be retained by the same TrustManager implementation as (2). For Go, auditing information seems it could be retained by tls.Config.VerifyPeerCertificate.

Those are pretty easy for an arbitrary application to implement, except for (2) with Go. So I don't think any changes need to be made to Java/Go gRPC.
